### ISSUE TITLE
Instance dependent definition before instance field in record causes internal error

### ISSUE THREAD

**Issue by mr-ohman** [agda/agda#5492]
The following code causes an internal error to occur in version 2.6.2 and Agda at commit e75750a55d8c45132f772eea341c7541f4989f04.
```agda
record R1 (A : Set) : Set where
  field
    fun : A → A

record R2 (A : Set) ⦃ r1A : R1 A ⦄ : Set where

instance
  r1 : ∀ {A} → R1 A
  r1 = record { fun = λ x → x }

record R3 : Set₁ where
  field
    A : Set

  r2A : R2 A
  r2A = record {}

  field
    ⦃ r1A ⦄ : R1 A
```
```
An internal error has occurred. Please report this as a bug.
Location of the error: __IMPOSSIBLE_VERBOSE__, called at src/full/Agda/TypeChecking/Monad/Signature.hs:727:32 in Agda-2.6.3-D2V0PkTCBiXC7IJuWrP7mz:Agda.TypeChecking.Monad.Signature
```
The error still remains if the type of `r2A` is changed to `R2 _`, however letting `r2A` have the type `R2 B` for some `B : Set` in `R3` makes the internal error go away. My guess is that `r1A` is somehow available for instance resolution in `r2A` in one part of the Agda implementation, which conflicts with the definition and field order.

**Comment by andreasabel** [comment]
I suppose your hunch is correct, turning on debug printing gives this:
```
...
{ solving constraint
[] Resolve instance argument _r1A_11 : R1 A No candidates yet
The type of the FindInstance constraint isn't known, trying to find it again.
Found instance type head:  R1
Getting candidates from context (r : R3)
Printing debug message tc.record.eta:20 failed due to error:
  /Users/abel/play/agda/bugs/Issue5492.agda:18,9-13
Panic: Unbound name: Issue5492.R3.r1A[0,28,40]@ModuleNameHash
1593967666935991171
when checking that A is a valid argument to a function of type
(A₁ : Set) ⦃ r1A : R1 A₁ ⦄ → Set
...
```
Hello #434, good to see you again!

**Comment by andreasabel** [comment]
This works with Agda 2.5.4 and 2.6.0, fails with 2.6.1.

**Comment by andreasabel** [comment]
Bisection says that b3aec6fa44a5ef077c30e563208c2274ef39e4d4 is the first bad commit.
Looking at that commit, it seems that it surfaces the problem rather than creating it...

**Comment by andreasabel** [comment]
If I replace the internal error by a `patternViolation` exception, we get an unsolved instance constraint:
```
Failed to solve the following constraints:
  Resolve instance argument _r1A_11 : R1 A
  Candidates
    Issue5492.R3.r1A : R1 A
    r1 : {A = A₁ : Set} → R1 A₁
    (stuck)
```
This may be better to work with.

I do not really understand how this commit can be introducing the internal error.  It is raised in `modalityOfConst`, but before the commit `relOfConst` should have raised it as well...

**Comment by jespercockx** [comment]
It seems this has been fixed in 2.6.2.1, so I'm closing this issue.
